src/tbb/cgroup_info.h

```diff
@@ -0,0 +1,308 @@
+                path_start = line + 3; // cgroup v2 unified path
+                relative_path = paths_cache.v2_relative_path;
+            } else {
+                // cgroups v1 allows comount multiple controllers against the same hierarchy
+                path_start = look_for_cpu_controller_path(line, last_char);
+                relative_path = paths_cache.v1_relative_path;
+            }
+
+            if (path_start) {
+                std::strncpy(relative_path, path_start, rel_path_size);
+                relative_path[rel_path_size - 1] = '\0'; // Ensure null-termination after copy
+
+                char* new_line = std::strrchr(relative_path, '\n');
+                if (new_line)
+                    *new_line = '\0'; // Ensure no new line at the end of the path is copied
+            }
+        }
+    }
+
+    static bool try_read_cgroup_v1_num_cpus_from(const char* dir, int& num_cpus) {
+        char path[PATH_MAX] = {0};
+        if (std::snprintf(path, PATH_MAX, "%s/cpu.cfs_quota_us", dir) < 0)
+            return false;       // Failed to create path
+
+        unique_file_t fd(std::fopen(path, "r"), &close_file);
+        if (!fd)
+            return false;
+
+        long long cpu_quota = 0;
+        if (std::fscanf(fd.get(), "%lld", &cpu_quota) != 1)
+            return false;
+
+        if (-1 == cpu_quota) {
+            num_cpus = unlimited_num_cpus; // -1 quota means maximum available CPUs
+            return true;
+        }
+
+        if (std::snprintf(path, PATH_MAX, "%s/cpu.cfs_period_us", dir) < 0)
+            return false;       // Failed to create path;
+        fd.reset(std::fopen(path, "r"));
+        if (!fd)
+            return false;
+
+        long long cpu_period = 0;
+        if (std::fscanf(fd.get(), "%lld", &cpu_period) != 1)
+            return false;
+
+        num_cpus = determine_num_cpus(cpu_quota, cpu_period);
+        return num_cpus != error_value; // Return true if valid number of CPUs was determined
+    }
+
+    static bool try_read_cgroup_v2_num_cpus_from(const char* dir, int& num_cpus) {
+        char path[PATH_MAX] = {0};
+        if (std::snprintf(path, PATH_MAX, "%s/cpu.max", dir) < 0)
+            return false;       // Failed to create path
+
+        unique_file_t fd(std::fopen(path, "r"), &close_file);
+        if (!fd)
+            return false;
+
+        long long cpu_period = 0;
+        char cpu_quota_str[16] = {0};
+        if (std::fscanf(fd.get(), "%15s %lld", cpu_quota_str, &cpu_period) != 2)
+            return false;
+
+        if (std::strncmp(cpu_quota_str, "max", 3) == 0) {
+            num_cpus = unlimited_num_cpus;  // "max" means no CPU constraint
+            return true;
+        }
+
+        errno = 0; // Reset errno before strtoll
+        char* str_end = nullptr;
+        long long cpu_quota = std::strtoll(cpu_quota_str, &str_end, /*base*/ 10);
+        if (errno == ERANGE || str_end == cpu_quota_str)
+            return false;
```

Comment:
Should not we reset `errno` before exiting?

Reply:
I don't think that it is strictly necessary, because usage pattern of `errno` implies that before any call that sets `errno`, it is better to reset `errno` to 0.